### PR TITLE
Feat: Dynamically set app in dev mode

### PR DESCRIPTION
**Pull request by jonatanhanson**
## Explain the changes you’ve made

Created a solution for entering the dev environment even when the app is build for production.
Also moved dev features to a new screen that is shown only in "dev mode".

## Explain why these changes are made

This is needed by Apple and Google when they are reviewing the app.  

## Explain your solution

Added a new .env variable that defines a test personal number. If that specific personal number is entered when logging in the app is redirecting all requests from prod(AWS) to dev. You are still required to log in with Bankid but will enter the test-environment instead. 
To change and access the dev/prod mode dynamically I created a new context for this.

Because context(and other hooks) cannot be accessed in helper functions, I also needed  to save the "app env" value in AsyncStorage for easy accessing. This part feels is a little hacky but was the best I could come up with, any suggestions for other solutions is welcome. 

Created a new screen for dev features that is shown in dev mode and can be accessed from the Profile screen.

## How to test the changes?

1. Add these .env variables:
APP_ENV=production
MITTHELSINGBORG_IO=[prod or stage aws url]
MITTHELSINGBORG_IO_APIKEY=[some api key]
MITTHELSINGBORG_IO_DEV=[develop aws url]
MITTHELSINGBORG_IO_DEV_APIKEY=[some api key]
TEST_PERSONAL_NUMBER=[199803312389] (person you have in test-Bankid)
USE_BANKID=true

2. Rebuild the app
3. Try to log in with a personal number that is not your test-person. And the app should request your prod-url.
4. To test dev environment. Log in with your test-user and the app should request your dev-url.
5. The app is now in "dev mode" and if you click on Profile screen you should see a button with "Utvecklarfunktioner".

## Was this feature tested in the following environments?
- [] Storybook on a iOS device/simulator.
- [X] Building the Application on a iOS device/simulator.
- [] Building the Application on a Android device/simulator.

### CODE REVIEW

**Comment by helsingborgsstad** [comment]
Task linked: [CU-cpz3cj Implementera appåtkomst för Apple + Google](https://app.clickup.com/t/cpz3cj)

File: source/helpers/UrlHelper.js [hunk 1]
```diff
@@ -47,16 +48,21 @@ const encodeQueryData = (queryParams) => {
...
-  let queryParams = { apikey: env.MITTHELSINGBORG_IO_APIKEY || '' };
+export const buildServiceUrl = async (endpoint = '', params = {}) => {
+  const appEnv = await StorageService.getData(APP_ENV_KEY);
+  const devMode = appEnv === 'development';
+  const apiUrl = devMode ? env.MITTHELSINGBORG_IO_DEV : env.MITTHELSINGBORG_IO;
+  const apiKey = devMode ? env.MITTHELSINGBORG_IO_DEV_APIKEY : env.MITTHELSINGBORG_IO_APIKEY;
+
+  let queryParams = { apiKey };
   // Concatenate params
   queryParams = { ...params, ...queryParams };
```
Comment (jwinding) [inline source/helpers/UrlHelper.js:59]: This looks weird, why not just 
```suggestion
const queryParams = { ...params, apiKey };
```

File: source/store/AppContext.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import PropTypes from 'prop-types';
+import React, { useState } from 'react';
+import env from 'react-native-config';
+import StorageService, { APP_ENV_KEY } from '../services/StorageService';
+
+const AppContext = React.createContext();
+
+function AppProvider({ children }) {
+  const acceptedModes = ['development', 'production'];
+  const defaultMode = env.APP_ENV || acceptedModes[0];
```
Comment (jwinding) [inline source/store/AppContext.js:10]: Maybe check if env.APP_ENV is one of the accepted modes?
```suggestion
  const defaultMode =
    env.APP_ENV && acceptedModes.includes(env.APP_ENV) ? env.APP_ENV : acceptedModes[0];
```
Can't trust developers to set variables correctly, after all.

File: source/store/AppContext.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    }
+  };
+
+  const provider = {
+    mode,
+    handleSetMode: (newMode) => handleSetMode(newMode),
+    isDevMode: mode === 'development',
+  };
+
+  React.useEffect(() => {
```
Comment (jwinding) [inline source/store/AppContext.js:25]: Very minor, but we usually import useEffect, and you already import useState, so for style it's nicer to import useEffect and not write React. here.

**Review by jwinding — COMMENTED** [review]
Code looks good, had some minor comments. Also sad that the new files are not in typescript, but that's fine. However, I do have some issue with how to test it, but I'll try to figure it out.

**Review by jwinding — COMMENTED** [review]
I have now tested it (in a semi-mock way where both prod and dev points to the same aws environment, but what you gonna do), and can confirm that it works; when the personal number is equal to the test-pnr, you go into develop-mode and the 'different' address gets used, and you can access the developer options. 

So, barring my minor code suggestions, I approve.

File: source/helpers/UrlHelper.js (hunk 1 above)
Comment (jonatanhanson) [reply source/helpers/UrlHelper.js]: That was weird. Fixed!

File: source/store/AppContext.js (hunk 2 above)
Comment (jonatanhanson) [reply source/store/AppContext.js]: Nope we are not trustworthy. Fixed!

File: source/store/AppContext.js (hunk 3 above)
Comment (jonatanhanson) [reply source/store/AppContext.js]: Fixed!

**Review by jwinding — APPROVED** [review]

**Comment by jonatanhanson** [comment]
> Code looks good, had some minor comments. Also sad that the new files are not in typescript, but that's fine. However, I do have some issue with how to test it, but I'll try to figure it out.

I will try to learn Typescript after the release, feel like I don't have time to slow down right now. 😅

**Comment by jonatanhanson** [comment]
Made new fix: 
Add api-key to all requests from the ApiRequest helper to make sure we use the correct one depending on dev/prod mode.

**Review by TeddyAndersson — APPROVED** [review]
Went through the code and it looks fine, have not tested this yet. But if this works for both you @jonatanhanson and @jwinding i approve this PR.

**Comment by jonatanhanson** [comment]
> Went through the code and it looks fine, have not tested this yet. But if this works for both you @jonatanhanson and @jwinding i approve this PR.

Goodie, but it would be good if someone tested it again because i had to rebase and make some additional fixes after Jacobs review.

**Pull request merged**